tests/nn/test_module.py

```diff
@@ -129,6 +134,31 @@ def __init__(self):
     assert_equal(f.child2.a.detach(), torch.tensor(2.))
 
 
+def test_module_cache():
+    class Child(PyroModule):
+        def __init__(self, x):
+            super().__init__()
+            self.a = PyroParam(torch.tensor(x))
+
+    class Family(PyroModule):
+        def __init__(self):
+            super().__init__()
+            self.c = Child(1.0)
+
+        def forward(self):
+            return self.c.a
+
+    f = Family()
+    f.c = Child(3.)
+    assert_equal(f().detach(), torch.tensor(3.))
+
+    g = Family()
+    g.c = Child(4.)
+    assert_equal(g().detach(), torch.tensor(4.))
+    # FIXME: failing test
+    assert_equal(f().detach(), torch.tensor(3.))
```

Comment:
@fritzo I give up dealing with this failing test. :( I think that failing is the expected behavior here (because we only have 1 global ParamStoreDict for 2 global PyroModule instances). Some possible solutions are
+ Set `_pyro_name = id(self)` by default. This will create ugly names for param sites in a trace.
+ Allow users to set `_pyro_name` in __init__ method.

Reply:
Nvm, this can be resolved with `pyro.clear_param_store()` or combining these two PyroModules into one. The later one will resolve the conflict of `f`'s parameters and `g`'s parameters in ParamStore.